File: python/all-your-base/all_your_base.py
def rebase(input_base, digits, output_base):
    if input_base < 2:
        raise ValueError("input base must be >= 2")

    if output_base < 2:
        raise ValueError("output base must be >= 2")

    if not all([d > -1 and d < input_base for d in digits]):
        raise ValueError("all digits must satisfy 0 <= d < input base")

    base10 = sum([ input_base ** place * digit for place, digit in enumerate(reversed(digits))])

    if base10 == 0:
        output = [0]
    else:
        output = []
        remainder = base10
        while remainder > 0:
            digit = remainder % output_base
            remainder = remainder // output_base
            output.insert(0,digit)

    return output

File: lua/all-your-base/all-your-base.lua
local all_your_base = {}

all_your_base.convert = function(from_digits, from_base)
    assert(from_base > 1, "invalid input base")
    local result = 0
    local exponent = 0

    for i = #from_digits, 1, -1 do
        assert(from_digits[i] > -1, "negative digits are not allowed")
        assert(from_digits[i] < from_base, "digit out of range")

        result = result + (from_digits[i] * from_base ^ exponent)
        exponent = exponent + 1
    end

    return {
        to = function(to_base)
            assert(to_base > 1, "invalid output base")

            local output = {}
            if result == 0 then
                table.insert(output, 0)
            else
                local remainder = result

                while remainder > 0 do
                    local digit = remainder % to_base
                    remainder = remainder // to_base
                    table.insert(output, 1, digit)
                end
            end

            return output
        end
    }
end

return all_your_base
